src/catalog/src/durable/impls/persist.rs

```diff
@@ -76,6 +78,8 @@ enum Mode {
 /// can expire its leases. If/when rust gets AsyncDrop, this will be done automatically.
 #[derive(Debug)]
 pub struct PersistHandle {
+    /// Since handle to control compaction.
+    since_handle: SinceHandle<StateUpdateKindBinary, (), Timestamp, Diff, i64>,
```

Comment:
Do we need to add anything to `compare_and_append` on line 455? That's where we currently call `self.read_handle.downgrade_since`.

Reply:
Possible! The `Subscribe` downgrades the since internally, which is why it's useful to downgrade the handle there.

I see `compare_and_append` is mostly used in debug code and in the initial fence of old writers. I don't feel particularly motivated in either direction by this. @danhhz has probably then most context on both sides of this... any opinion?

Reply:
> It's also used in `commit_transaction`...

Ah, I was talking about `PersistHandle::compare_and_append` (which does the since downgrade) but `PersistCatalogState::compare_and_append` is the one called by `PersistCatalogState::commit_transaction` (which does not).

> Yeah, I think go ahead and add it in the compare_and_append.

🙇

Reply:
Okay in that case, it sounds like we probably want to put a `allow_downgrade` call after all of the `compare_and_append` calls

Reply:
I appreciate this thread is getting a bit long already, but I'm not sure that makes sense to me, and this feels like the sort of reasoning I'm going to need to repeat a bunch if I'm actually going to tackle the lease thing. (Hypothetically!) So bear with me if you can!

If we allow-downgrade in compare-and-append, IIUC, a running catalog will be downgrading in two places:
- After an append, we'll immediately downgrade the since. At this point the subscribe is potentially "dangling" - the critical handle is possibly later than the since's current position.
- After the subscribe reads the data back, we'll downgrade the since again. We already downgraded the since after the CaA, though, so this is a noop.

If we're just trying to fix https://github.com/MaterializeInc/materialize/issues/22789, I think downgrading in either location would be fine, since we only really care about the since being held back if all leases expire. But if we're trying to have all active reads protected by the critical handle, skipping the update during the CaA and doing it as the subscribe makes progress seems like the right move.

(And as a meta point - this is too tricky of reasoning to do often. If this isn't just a one-off complexity since everything else uses the controller's tracking, we'll need to do better here.)

Reply:
I think I'm confused because the motivations for this PR are somewhat unclear to me. The actual guarantee needed by the current catalog code is just that the since doesn't pass the upper. Even accounting for unwanted read lease timeouts, I'm not sure I understand how the current code would lead to a bug!

That leaves setting up future work to remove read leases as a motivation. But that's confusing to me because Subscribes need both a since and a seqno hold. The idea we had to remove read leases involved figuring out some sort of mapping from progress ->  the first seqno that might still be relevant -> a since. This is plugging "progress-1" as the new since as the protection for the Subscribe, but that's not sufficient in a leaseless reader world. So, I'm having a hard time understanding the distinction you're making here.

Reply:
> I think I'm confused because the motivations for this PR are somewhat unclear to me.

Cool!

> The actual guarantee needed by the current catalog code is just that the since doesn't pass the upper. Even accounting for unwanted read lease timeouts, I'm not sure I understand how the current code would lead to a bug!

If you asked me what the since of a Persist shard was today, I'd say "the meet of all the critical and leased handles" and then mutter something about handle expiry and other exceptions. If that simple definition were true, the catalog's since would go to `{}` if there was no active lease, which would be bad. Are our exceptions enough to protect the catalog? Are there cases where this would already cause a bug? I'd have to think about it! But AFAICT this is the only place where the answer to that question is complicated, since every other shard has a critical handle in place... seems better to use one here too, especially since it is also very easy to do.

And yep, if that's the only motive then holding the since behind the upper is fine. I'd be happy to tweak this PR enforce only that guarantee if that's more compelling to you!

Reply:
> That leaves setting up future work to remove read leases as a motivation.

The main thought here was that:
- At least some of the long-term ideas we've talked about involve the since hold being kept behind the active reads. (And in general, we manage the since on behalf of current and future readers, not the writer.)
- That is easy to do, and more or less replicates how the lease was already managing the since, so it was unlikely to cause new problems.

So, seemed nice and vaguely forward-looking, but there's not a particularly crisp vision for it. If you'd rather not worry about it and stick to the more concretely motivated writer-lagging version for now, then that's good with me.

Reply:
Okay, that all makes sense! I have some responses, but it feels a bit to me like we're past the point of efficient async communication on this, so I put it on the persist weekly agenda to hash out the details and get this merged